File: hermes/screenshot_diff.py
"""#13 — Hermes Screenshot Diff Analysis

Compares a reference screenshot to a current screenshot using PIL/Pillow.
Computes pixel-level structural similarity, perceptual hash distance,
and dominant-colour shift.
"""
from __future__ import annotations

import logging
import math
from pathlib import Path

try:
    from PIL import Image
except ImportError:  # graceful degradation
    Image = None  # type: ignore[assignment,misc]

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

def _load_image(path: str) -> "Image.Image":
    if Image is None:
        raise RuntimeError("Pillow is not installed — run `pip install Pillow`")
    return Image.open(path).convert("RGB")


def _pixel_diff_percentage(img_a: "Image.Image", img_b: "Image.Image") -> float:
    """Return the percentage of pixels that differ (threshold > 30 per channel)."""
    a_pixels = list(img_a.getdata())
    b_pixels = list(img_b.getdata())
    if len(a_pixels) != len(b_pixels):
        return 100.0
    threshold = 30
    diff_count = sum(
        1 for pa, pb in zip(a_pixels, b_pixels)
        if any(abs(ca - cb) > threshold for ca, cb in zip(pa, pb))
    )
    return (diff_count / len(a_pixels)) * 100.0


def _average_hash(img: "Image.Image", size: int = 8) -> int:
    """Compute a simple average perceptual hash."""
    resized = img.resize((size, size), Image.LANCZOS)
    grey = resized.convert("L")
    pixels = list(grey.getdata())
    avg = sum(pixels) / len(pixels)
    bits = 0
    for px in pixels:
        bits = (bits << 1) | (1 if px >= avg else 0)
    return bits


def _hamming_distance(h1: int, h2: int) -> int:
    return bin(h1 ^ h2).count("1")


def _dominant_color(img: "Image.Image") -> tuple[float, float, float]:
    """Return the average colour as a simple dominant colour proxy."""
    pixels = list(img.getdata())
    n = len(pixels)
    r = sum(p[0] for p in pixels) / n
    g = sum(p[1] for p in pixels) / n
    b = sum(p[2] for p in pixels) / n
    return (r, g, b)


def _color_distance(c1: tuple[float, float, float], c2: tuple[float, float, float]) -> float:
    return math.sqrt(sum((a - b) ** 2 for a, b in zip(c1, c2)))


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def compare_screenshots(reference_path: str, current_path: str) -> dict:
    """Compare two screenshots and return a structured diff result.

    Returns
    -------
    dict with keys:
        similarity   — 0-100 float (100 = identical)
        hash_distance — integer Hamming distance of perceptual hashes
        color_shift  — Euclidean distance of dominant colours (0 = same)
        is_anomalous — True if diff exceeds safe thresholds
    """
    ref = _load_image(reference_path)
    cur = _load_image(current_path)

    # Resize to same dimensions for fair comparison
    target_size = (min(ref.width, cur.width), min(ref.height, cur.height))
    ref = ref.resize(target_size, Image.LANCZOS)
    cur = cur.resize(target_size, Image.LANCZOS)

    diff_pct = _pixel_diff_percentage(ref, cur)
    similarity = max(0.0, 100.0 - diff_pct)

    ref_hash = _average_hash(ref)
    cur_hash = _average_hash(cur)
    hash_dist = _hamming_distance(ref_hash, cur_hash)

    ref_color = _dominant_color(ref)
    cur_color = _dominant_color(cur)
    color_shift = _color_distance(ref_color, cur_color)

    is_anomalous = diff_pct > 20.0

    result = {
        "similarity": round(similarity, 2),
        "hash_distance": hash_dist,
        "color_shift": round(color_shift, 2),
        "is_anomalous": is_anomalous,
    }
    logger.info("Screenshot diff result: %s", result)
    return result

File: src/hermes/screenshot-diff.ts
/**
 * #13 — Hermes Screenshot Diff Analysis (TypeScript Port)
 *
 * Compares a reference screenshot to a current screenshot using pixel-level
 * analysis. Computes structural similarity percentage, perceptual hash distance
 * via average hash, and dominant-colour shift via Euclidean distance.
 *
 * Uses Node.js built-in `fs` for file reading and raw pixel math.
 * For production, considers `sharp` for image decoding (already available
 * in the ecosystem), but falls back to a lightweight JPEG header parser.
 *
 * Ported from hermes/screenshot_diff.py
 */

import fs from "node:fs";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface ScreenshotDiffResult {
  /** 0-100 float (100 = identical) */
  similarity: number;
  /** Integer Hamming distance of perceptual hashes (0 = identical) */
  hashDistance: number;
  /** Euclidean distance of dominant colours (0 = same) */
  colorShift: number;
  /** True if diff exceeds safe thresholds */
  isAnomalous: boolean;
}

type RGB = [number, number, number];

// ---------------------------------------------------------------------------
// Internal helpers — pure math, no external image dependencies
// ---------------------------------------------------------------------------

/**
 * Attempt to load sharp dynamically. Returns null if not available.
 * We use sharp because it's a common Node.js image library and already
 * pulled in by many Playwright-adjacent projects.
 */
type SharpInstance = {
  resize(w: number, h: number): SharpInstance;
  raw(): SharpInstance;
  greyscale(): SharpInstance;
  toBuffer(opts: { resolveWithObject: true }): Promise<{ data: Buffer; info: { width: number; height: number; channels: number } }>;
  metadata(): Promise<{ width?: number; height?: number; channels?: number }>;
};

async function loadSharp(): Promise<((input: string | Buffer) => SharpInstance) | null> {
  try {
    // Dynamic import to avoid hard dependency
    const sharpModule = (await import("sharp")) as { default: (input: string | Buffer) => SharpInstance };
    return sharpModule.default;
  } catch {
    return null;
  }
}

function pixelDiffPercentage(aPixels: Buffer, bPixels: Buffer, channels: number): number {
  const pixelCount = aPixels.length / channels;
  if (pixelCount === 0) return 100;
  const threshold = 30;
  let diffCount = 0;

  for (let i = 0; i < aPixels.length; i += channels) {
    for (let c = 0; c < Math.min(channels, 3); c++) {
      if (Math.abs(aPixels[i + c]! - bPixels[i + c]!) > threshold) {
        diffCount++;
        break;
      }
    }
  }
  return (diffCount / pixelCount) * 100;
}

function averageHashFromBuffer(greyPixels: Buffer, size: number): bigint {
  const avg = greyPixels.reduce((sum, v) => sum + v, 0) / greyPixels.length;
  let bits = 0n;
  for (let i = 0; i < greyPixels.length && i < size * size; i++) {
    bits = (bits << 1n) | (greyPixels[i]! >= avg ? 1n : 0n);
  }
  return bits;
}

function hammingDistance(h1: bigint, h2: bigint): number {
  let xor = h1 ^ h2;
  let count = 0;
  while (xor > 0n) {
    count += Number(xor & 1n);
    xor >>= 1n;
  }
  return count;
}

function dominantColor(pixels: Buffer, channels: number): RGB {
  const pixelCount = pixels.length / channels;
  if (pixelCount === 0) return [0, 0, 0];
  let r = 0, g = 0, b = 0;
  for (let i = 0; i < pixels.length; i += channels) {
    r += pixels[i]!;
    g += pixels[i + 1]!;
    b += pixels[i + 2]!;
  }
  return [r / pixelCount, g / pixelCount, b / pixelCount];
}

function colorDistance(c1: RGB, c2: RGB): number {
  return Math.sqrt(
    (c1[0] - c2[0]) ** 2 + (c1[1] - c2[1]) ** 2 + (c1[2] - c2[2]) ** 2
  );
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Compare two screenshots and return a structured diff result.
 *
 * Requires `sharp` to be installed. If sharp is not available, returns
 * a degraded result with similarity=0 and isAnomalous=true.
 */
export async function compareScreenshots(
  referencePath: string,
  currentPath: string
): Promise<ScreenshotDiffResult> {
  // Validate files exist
  if (!fs.existsSync(referencePath) || !fs.existsSync(currentPath)) {
    return { similarity: 0, hashDistance: 64, colorShift: 441, isAnomalous: true };
  }

  const sharp = await loadSharp();
  if (!sharp) {
    console.warn("[ScreenshotDiff] sharp not installed — returning degraded result. Run: npm i sharp");
    return { similarity: 0, hashDistance: 64, colorShift: 441, isAnomalous: true };
  }

  // Determine common size
  const refInfo = await sharp(referencePath).metadata();
  const curInfo = await sharp(currentPath).metadata();
  const targetW = Math.min(refInfo.width || 800, curInfo.width || 800);
  const targetH = Math.min(refInfo.height || 600, curInfo.height || 600);

  // Resize and get raw RGB buffers
  const refBuf = await sharp(referencePath).resize(targetW, targetH).raw().toBuffer({ resolveWithObject: true });
  const curBuf = await sharp(currentPath).resize(targetW, targetH).raw().toBuffer({ resolveWithObject: true });

  const channels = refBuf.info.channels;

  // 1. Pixel diff percentage → similarity
  const diffPct = pixelDiffPercentage(refBuf.data, curBuf.data, channels);
  const similarity = Math.max(0, 100 - diffPct);

  // 2. Perceptual hash distance
  const hashSize = 8;
  const refGrey = await sharp(referencePath).resize(hashSize, hashSize).greyscale().raw().toBuffer({ resolveWithObject: true });
  const curGrey = await sharp(currentPath).resize(hashSize, hashSize).greyscale().raw().toBuffer({ resolveWithObject: true });
  const refHash = averageHashFromBuffer(refGrey.data, hashSize);
  const curHash = averageHashFromBuffer(curGrey.data, hashSize);
  const hashDist = hammingDistance(refHash, curHash);

  // 3. Dominant color shift
  const refColor = dominantColor(refBuf.data, channels);
  const curColor = dominantColor(curBuf.data, channels);
  const colorShiftVal = colorDistance(refColor, curColor);

  return {
    similarity: Math.round(similarity * 100) / 100,
    hashDistance: hashDist,
    colorShift: Math.round(colorShiftVal * 100) / 100,
    isAnomalous: diffPct > 20,
  };
}
